aten/src/ATen/native/Histogram.cpp

```diff
@@ -0,0 +1,133 @@
+#include <ATen/ATen.h>
+#include <ATen/Dispatch.h>
+#include <ATen/NativeFunctions.h>
+
+#include <ATen/native/Histogram.h>
+
+#include <tuple>
+
+namespace at { namespace native {
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+DEFINE_DISPATCH(histogram_stub);
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+DEFINE_DISPATCH(histogram_linear_stub);
+
+namespace {
+
+inline void histogram_pre_check(const Tensor& input, const Tensor& bins, const c10::optional<Tensor>& weight,
+    bool density, const Tensor& hist, const Tensor& bin_edges) {
+  TORCH_CHECK(bins.device() == input.device(), "bins and input value tensors should have same device type, ",
+    "but we got bins tensor device type ", bins.device(), " and input value tensor device type ", input.device());
+
+  TORCH_CHECK(bins.dim() == 1, "bins tensor should have dimension 1, but got ", bins.dim(), " dimension");
+
+  TORCH_CHECK(hist.dim() == 1, "hist tensor should have dimension 1, but got ", hist.dim(), " dimension");
+
+  TORCH_CHECK(bin_edges.dim() == 1, "bin_edges tensor should have dimension 1, but got ", bin_edges.dim(), " dimension");
+
+  TORCH_CHECK(bins.numel() > 0, "bins tensor should have at least 1 element, but has 0");
+
+  if (weight.has_value()) {
+    auto weight_sizes = weight.value().sizes();
+    TORCH_CHECK(input.sizes() == weight.value().sizes(), "if weight tensor is provided, input value tensor and weight tensor",
+        " should have the same shape, but we got input(", input.sizes(), ") and weight(", weight_sizes, ")");
+
+    ScalarType weight_dtype = weight.value().scalar_type();
+    TORCH_CHECK(weight_dtype == ScalarType::Float || weight_dtype == ScalarType::Double,
+        "weight tensor's dtype must be Float or Double, but we got weight tensor's type ", weight_dtype);
+  }
+
+  TORCH_CHECK(hist.is_contiguous(), "hist output tensor must be contiguous");
+
+  TORCH_CHECK(bin_edges.is_contiguous(), "bin_edges output tensor must be contiguous");
+
+  ScalarType hist_dtype = hist.scalar_type();
+  if (density) {
+    TORCH_CHECK(hist_dtype == ScalarType::Double, "hist tensor's dtype is wrong, it must be Double if density flag is True, "
+        "but we got hist tensor's type ", hist_dtype, " and density flag is ", (density ? "True" : "False"));
+  } else if (weight.has_value()) {
+    TORCH_CHECK(hist_dtype == ScalarType::Double, "hist tensor's dtype is wrong, it must be Double if weight is provided, "
+        "but we got hist tensor's type ", hist_dtype);
+  } else {
+    TORCH_CHECK(hist_dtype == ScalarType::Long, "hist tensor's dtype is wrong, it should be long in the unweighted case ",
+        "but we got hist tensor's type ", hist_dtype);
+  }
+}
+
+ScalarType hist_scalar_type(const c10::optional<Tensor>& weight, bool density) {
```

Comment:
Let's talk about the operation's type promotion (or lack thereof).

What are the options for how types are handled?

We could require, for example, that every tensor be a floating point tensor of the same dtype, and only support the float and double dtypes, that might be simple

But what types does histc support? If we're removing the TH histc impl we should be sure to implement all of those, too

Do we need special kernel instantiations to handle integer inputs tensors?